### ISSUE TITLE
Advanced Edge Authentication (Username & Password)

### ISSUE THREAD

**Issue by Logan007** [ntop/n2n#670]
So far, edges authenticate by knowing the pre-shared key and the right community name. They prove their knowledge by correctly applying it to payload encryption and / or header encryption respectively.

But what happens if an edge gets stolen, e.g. in case of mobile devices? To get the thief out of the network the key on every single edge would need to be changed.

We were thinking of some more advanced edge authentication scheme. I do not want to go fully "public key" yet as it brings up a lot of PKI related questions which I do not want to load on n2n for now.

So, we internally discussed a scheme which I will put into the following post – very lengthy! **TLDR**: You can remove an edge from the network by deleting one line from a text file residing at the supernode instead of manually re-keying all your other edges. It is based on user-password / public key authentication.

I would be most interested in your opinion, especially if we shall integrate it into upcoming 3.0.

**Comment by Logan007** [comment]
# Advanced Edge Authentication

## Requirements

Each edge needs to have a unique authentication string, provided via an additional command line parameter `-J <string>`. Together with the already present `-I <string>` identificator which, as a new requirement, must be unique as per community, it is used for authentication.

The supernode needs a list of all valid `I:J` pairs. One could think of them as username (I) and password (J). Inside the community.list file, these `I:J` pairs could follow the line with the community name:

```
<community name>
* <identifier 1> <j auth value 1>
* <identifier 2> <j auth value 2>
* <identifier 3> <j auth value 3>
...
```

A special character at the beginning indicates I:J pairs. A sane regular expression cannot start neither with an asterisk `*` nor with a closing bracket `]`, so those lines are distinguishable.

## Basic Idea

A hash value of the `-J <string>`-provided string is then used as key _Kj_ only known to this edge and the supernode.

Upon registration with the supernode, the edge symmetrically encrypts a random number _r_ with key _Kj_. The supernode finds the local _J_ value by index _I_,it is sent inside each REGISTER_SUPER, decrpyts it and adds the current community key _Kc_ to _r_. The sum gets encrypted again with key _Kj_ and is sent back with the REGISTER_SUPER_ACK. The edge is able to decrypt it and subtract _r_ again to obtain the current communiry key _Kc_. This is the key used for header encryption of all packets. Only exception: REGISTER_SUPER and REGISTER_SUPER_ACK still get encrypted using the hash of the community name as key so registration can always happen.

So, this scheme requires enabled header encryption and therefore a fixed community name.

If no corresponding I:J pair is found at the supernode, the supernode does not answer, the edge will not be able to participate because it does not know the current community key _Kc_.

## Details

### How to Enable

This mode is enabled on a per community basis. The edge enables this mode by the presence of `-J <string>`. The supernode knows about as soon as a community has I:J pairs listed.

### Cryptography

The current header encryption does not protect the payload which is protected by the `-k` key which also is known to a stolen edge. That's why we need to ensure that packet can only be decrypted knowing current _Kc_. Using the stream ciphers SPECK or ChaCha20, no part of the packet content can be decrypted without the knowledge of the IV (this is not true for the block ciphers in CBC-CTS mode). Thus, the header encryption is extended to the beginning of the data packet where the payload IV resides and only stream ciphers are allowed.

The block cipher schemes optionally could be extended to be encapsulated by some additional quick Even Mansour scheme using the payload IV (hash values thereof) as keys. Then, the payload cannot be decrypted without having the restored the IV. These schemes could optionally be added as their own implementations or added to the existing Twofish and AES schemes.

The current community key _Kc_ is described further below in detail.

### Remove a Stolen Edge from the List

If an edge gets stolen or needs to be dis-legitimated for some reason, it just gets removed from the `community.list`. Comments in that file along with human readable `-I`-supplied identifiers might be helpful to identify the proper entrylist. In such an event, a re-keying is required, a new community _Kc_ is generated and used for packet encryption. Such a re-keying event could be triggered by a command issued to the management port console after having deleted the I:J entry.

### List of Edges in Case of Federation

As of now, REGISTER_SUPER packets from edges get forwarded to all other supernodes so they have a chance to disagree on registration (sending REGISTER_SUPER_NAK to originating supernode) and gather information for packet forwarding. In case of positive authentication, there is no REGISTER_SUPER_ACK sent from the other supernodes to the original supernode.

Hence, in case of advanced edge auth, the orginal supernode needs to delay its answer to the edge's REGISTER_SUPER request, for let's say three seconds, to be sure no REGISTER_SUPER_NAK has arrived from the other nodes. If it receives such a NAK or has not find valid credentials for the edge either, it declines registration.

Therefore, an edge will be denied if it not known by all supernodes. By consequence, the lists needs to be synchronized between the supernodes for which external tools such as syncthing would be recommended.

On the one hand, deleting an entry on an online supernode should be enough to keep the corresponding edge away. Two re-registration periods later, a deleted edge will not be able to connect to other edges anymore. On the other hand, an edge only gets access if its entry is present at all online supernodes.

The three second delay can be implemented with a hash list of timed events which gets checked in the main loop for still-to-be-done events. The registration timing needs adjustment for the delay does not get overtaken by the edge's next try to register. Perhaps it could be switched to 5 second re-registration interval (5 > 3) and in turn using only 2 re-registration attempts before switching to the next supernode.

### Community Key

The community _Kc_ is calculated by

```
HASH (  HASH(Community Name)  XOR  HASH(Federation Name)  XOR  Timestamp  )
```

This way, the key depends on the community name, cannot be calculated by the edges due to the secret federation name, and can be calculated by other supernodes as well due the special time stamp. The federation name must be kept secret.

The time stamp is the 32-bit UTC time. The time stamp from key generation time is globally stored at the supernode (sss->key_time). That way, federated supernodes can exchange information on the used key time. If two supernodes serving the same community meet for the first time they can determine which supernode has the older key and needs to update, that is to re-key to the younger one, just find max(time1, time2). Remember that supernodes' federation communication is header encrypted so their clocks cannot differ more than 16 seconds by default anyway.

This time information can be exchanged in the auth-field of the REGISTER_SUPER and REGISTER_SUPER_ACK between federated super nodes because they do not have to use the auth field the way the edges do.

To avoid the need of handling different key times for different communities, the key-time is global and valid for all communities. So, a re-key event changes the keys of all communities which is no issue because of keeping the previous key valid for a certain period of time.

### Re-Keying Event

The re-keying should be announced by a new message type REKEY between the supernodes. It shall include the now valid key_time. Supernodes compare the new keytime and re-calculate their community keys if required.

This comprises the manually triggered re-keying as well as a re-keying performed on the basis of differing key_times found from a new federated supernode registration. To prevent a special scenario with differing clocks (even +/- 16 sec only: all key times synced to ...50; at time ...100, sn1 with clock +-0 re-keys for whatever reason, lifting federation up to key time ...100; at time 105 sn2 with clock -10 re-keys to new keytime 95?! the other supernodes will not accept the actually newer key with an older time index), it is recommended to use 2 * 16 seconds after the last re-key (timer goes back to 32 if new re-key meanwhile) as re-key time.

To avoid edge-cases in overlapping key-handling and times of often happening re-keying events, the edges could be notified as well by a plain REKEY message without further information. It would cause the edge to re-register immediately.

### Conclusion

This requires some implementation work touching various places in the code.

The usage of this feature requires to keep the federation name secret (should be known to administrators of supernodes only anyway) and to manage the auth strings at the edge as well as their lists at the supernodes. Supernodes clocks should be in sync as far as possible.

This is probably the last things we can implement with the help of symmetric ciphers before the code gets unreadable. For future versions, perhaps 4.x, we might want to switch to asymmetric cryptography using certificates. On the other hand, that would require to add PKI-like handling including revocation mechanism – in a user-friendly way.

**Comment by Logan007** [comment]
I am not too happy with having the J-values, the "passwords", in plain text at the supernodes. But every challenge-response with salted password hash storage at the server-side I am aware of involves at least 4 messages as opposed to our 2 (REGISTER_SUPER and REGISTER_SUPER_ACK).

Some very "mild" public key cryptography could help. Maybe Curve25519 to keep a private J at the edge (salted hash of user specified password) and the corresponding public J at the supernode?

Any thoughts? Do we want this feature anyway?

By the way, this is related to #30.

**Comment by tlsalex** [comment]
Two factor authentication or PKI is so good (maybe for enterprise ready), but only one man  work for this project, that work is too much.

**Comment by Logan007** [comment]
> Two factor authentication or PKI ... too much

I agree that those definitely are out of reach for 3.0 ...

But per-edge username & password should be achievable. The question is just: Do we need and / or want it? Let me know!

**Comment by xaverruss** [comment]
If a username/password combination identifies a single node, that would lay the groundwork for static IP and MAC assignments on the supernode side and prevent the "steal" of an IP a currently offline device would have. 

I have used zerotier before I found N2N and this central static assignment and per node activation is something I miss.
Also, if this gets added as API commands one could build a webfront with usermanagement and ACL for community based edge add and removal, much like a zerotier controller.

If this static assignment of IP/MAC would be optional, automatic mass deployments on Cloud Providers would simply get 
an edge group key and password combination. ... replaceing the key on a per node or per autodeploy script base is definetly easier than replacing it on all nodes in the community.

**Comment by Logan007** [comment]
@xaverruss, thank you for your detailed comment!

> groundwork for static IP and MAC assignments

You are right. If we start to add per-edge lines to a config file, I was thinking about the `community.list` file, we could also add some more (optional) fields for MAC and IP address. I do not want to envision all the format sanity checks yet… :wink:

The current protocol only allows for IP address assignment by supernode (omit `-a`), so another preparatory change would be required. Preparatory because MAC assignment would most probably follow at some later point in time. If there is a need for a MAC assignment feature, this must be prepared before a 3.0 release. Aren't IP addresses sufficient while keeping MAC addresses random?

> prevent the "steal" of an IP

Concerning the IP address assignment, you currently get distinct and unique IP addresses if you carefully chose the username (`-I <...>`). A tool could help with it. On the weak side, it relies on edge's cooperation in that it provides the correct string.

> per node activation

Yes, that is exactly the driving idea behind it as well as the opportunity for de-activation.

> one could build a webfront with usermanagement

A wonderful idea! It would aim at managing the `.conf`-file mentioned above in a Unix `passwd` / `shadow` fashion. I am not so sure about APIs...

> simply get an edge group key and password combination
> replaceing the key on a per node or per autodeploy script base

I am not sure if I get you correctly. Do you suggest that edges should receive all their parameters from the supernodes including the payload key just by identifying themselves with username and password? That sounds like a very interesting idea! I just fear that it goes against the current approach which makes sure that especially the supernode does not know the payload key.

If you talk about the "session" key, it can be replaced and distributed among the "good" edges only hot-plug style as outlined in the concept above.

**Comment by xaverruss** [comment]
First I thought that you still would have the current encrytion key and share that among the edges without letting the supernode know it. 
You still could deactivate the edge nodes from the supernode by removing the edge authentication.
But this would lead to possible mitm attacks against the remaining nodes once one got ... compromised.

As said in the wireguard issure you mentioned, having a per node public/private keypair instead of the community preshared key would increase the secutrity to massive extend.

I though about some points:
Supernode:
- relays traffic if nessesary just as currently the case
- Stores community Information like name, possibly encryption chipper used
- Stores an edge id, could be random generated on the edge with "edge --createid"
   this ID identifies the client and allows access to the community, and would just be a long string that both supernode and    
   edge have in cleartext
- You then could assign an IP (possibly MAC if needed) to an edge node on the supernode side.
   having a central IP assignment authority makes more sense to me than having each edge do this by themself.
   - either you are in controll over the supernode and all the edge nodes, and have to keep a list of 
      all already assigned IPs anyway 
   - or you are not in controll over the edge nodes and do not want the users of those to assin an IP at will.
   - if you are on the edgeside and cant controll the supernode, you also might not want to grab an IP a other user already 
      has
- if you ommit the IP assignment (still on the supernode) the edge would get an ip in a space specified like 
   currently the case. 
   This IP assignment coud be saved on the supernode, so it gets reassigned to the edge if it went 
   offline, just like you meantioned "edge -I" would work.
- having an option for TTL for either the random assigned IPs or even the the edge IDs could be usefull too.
   for example if you have a cloud deployment on AWS Spot instances and the instances are permanently gone after a short time, you do not want to manually free the IPs in the space.
- on the other hand, an edgeID ... but instead of assigning an IP, one could add a "group" tag to the entry.
   or call it groupID alltogether to better differenciate the two 
   if this one is set, multiple edges with this groupID could join, every one getting its own IP in the range, but without a 
   stickyness factor, so once the edge is gone the addess assignment is gone, 
   if the node reconnects it simply gets a new ip for its connection lifespan
   This would also remove the need for the TTL I just came up with, I wanted to delete the TTL point at first, but it might 
   be a usefull thought to keep arround.
-  for every edge using a group ID the supernode should keep a seperate information table, including the public key ofc


> I just fear that it goes against the current approach which makes sure that especially the supernode does not know the payload key.

I havent though about the supernode beeing in untrusted hands before, but makes sense that this might be the case for some.

If the encription key would moved from preshared key to edge specific public/private keypairs the supernode would have zero knowlage about what is send between the nodes, even if it relays the traffic. 

I am sorry if I have mentioned ideas or stuff others already have, just wanted to write a (at least in my head) complete toughthrough down.

Also I have a question:
Does the supernode currently safe the random assigned IPs to disk somewhere ? or is it all in memory and once I restart  the supernode those are gone ?

Adding to this, I think having a real database on disk of some sort, ramcached ofc. Might bring some consistency to the supernode. 
If this would be implemented, everything in one long community file might not cut it for larger deployments, a mysql-lite  database with multiple tables linking to one an another would. this also could add to better multi supernode setup, because many databases already have somesort of replication or syncronisation, which could be handelt over something like the current federation network.


PS:
If the supernode holds information about the version the edgenode is running, this would allow for a better upgrading path in between releases as options could change or added and fail to be interpreted by other edges if the version is not the same.
Adding to this a automatic update system could be a usefull intigration later on.
If the supernode are unter your own controll, it hosting an up to date repository and the edges pulling from those and compiling with also containerised building dependencys ... but this idea is a BIG IF, and will take some carefull thinking as it could beak massive productive deployments in one go

**Comment by Logan007** [comment]
These are a lot of interesting ideas!

> But this would lead to possible mitm attacks against the remaining nodes once one got ... compromised.

That is why header encryption (`-H`)  shall get a dynamic session key and be enhanced to also (in addition to the still present payload encryption) protect the payload. Or, where exactly do you see the mitm attack potential?

>  per node public/private keypair instead of the community preshared key would increase the secutrity to massive extend

Well, at least it would make it easier to switch keys if compromised because you only need to switch the one compromised.

> If the encription key would moved from preshared key to edge specific public/private keypairs the supernode would have zero knowlage about what is send between the nodes, even if it relays the traffic.

Some issues however remain. Broadcasts require a community-wide key which shall be unknown to the supernode. And key sharing / common secret generation needs to be considered and integrted into the protocol.

Also, if we start using public keys, I want to do it somewhat right. That would include to use maybe not full certificates but at least signed keys allowing for verification.

 >  assign an IP (possibly MAC if needed)

We could add that to the `<user:"password">` entries at the supernode as option: `<user:"password":IP address: MAC address>` though I still am not sure about the MAC address.

> Supernode:

Yes, the supernode currently serves a lot of functions, I have [collected](https://github.com/ntop/n2n/issues/405#issuecomment-756893460) some of them. As discussed in that isuue #405, going full peer-to-peer is an option in the future. That's why I would prefer any supernode functionality not being dependant on its central position in the network.

> Does the supernode currently safe the random assigned IPs to disk somewhere ? or is it all in memory and once I restart the supernode those are gone ?

No, n2n itself does not write anything to the file system. The IP addresses to be assigned are not really random and calculated as follows:

First part (network part) = from net mask
Last part (host part) = from hash of edge identifier (`-I <identifier>`)

If that address turns out to already be assigned, the next higher or (afterwards) lower ones are tried.

This algorithm originates from a time when we only had one supernode and is not a distributed one. However, if the edeg identifier are properly chosen, i.e. no cutting hashes, this would also work fine. One could develop a small tool working on a list of already used names and indicating if a new given name would be okay, i.e. does not cut any other hash / IP address, maybe even propose an suffix to the new name tried  to make the hash unique again – just like when applying for a new gmail address.

> many databases already have somesort of replication or syncronisation

n2n aims at being very low footprint and I have heard from a lot of people running edges and supernodes on router hardware with low memory, disque as well as RAM, and CPU power definitely below the 1 GHz range. While n2n still is not the main task being executed on those devices, it only can consume some of their resources.

This is one of the reasons why n2n tries to stay away from external dependencies, even from openSSL which seems too heavy for some users; we integrated all the ciphers completely instead and left openSSL support optional.

So, as tempting it might be from a coder's perspective to leave the challenging part to other packets, aiming at slim code seems to be a current priority but leaves some own development potential to n2n.

> automatic update system

A very visionary idea and interesting approach you are suggesting. I wish somebody could implement it.

> just wanted to write a complete toughthrough down

Thank you for sharing your ideas. I am looking forward to further fruitful discussions. We probably soon will need to open some 4.0 discussion space... :wink:

**Comment by tlsalex** [comment]
#671 is much more about data encryption for 4.x
And here I want to give it a blueprint for authentication for 5.x -- Two factor authentication (combination of Username&Password and Certification,  or microsoft-like, google-like authenticator in smartphone), 

Bellow is the ways for authentication in the VPN application in enterprise that I ever used.
"Pulse Secure" use the combination of user's credential in Active Directory and  authentication code.
”OpenVPN“ use the combination of user's credential in Active Directory and certification.

**Comment by Logan007** [comment]
5.0 – @tlsalex, you are visionary!! :orange_heart:

I like the idea of making n2n seamlessly slip into more professional environments. I just hope that serving the corresponding interfaces does not take too much of its lightweight-ness.

**Comment by Logan007** [comment]
Okay, so let's follow this road. I will start implementation of user-password-style authentication to have a basis for further features and discussions. I hope that I will have it ready for 3.0.

Having such a feature in the 3.x series allows to gain experience and will show if and what kind of authentication we will require in 4.x – and beyond.

**Comment by Logan007** [comment]
A further requirement could be to make sure that an edge is connected to a legit supernode. Just imagine that an edge got stolen or compromised and was removed from the list of valid edges.

Evaluating the information an attacker got from a stolen edge, especially the community name, an attacker can setup a malicious supernode. If then, in addition, the attacker can trick a legit edge to connect to that manipulated supernode, the attacker can communicate with the legit edge.

To avoid it, we need to make sure that the dynamic header key originates from a legit supernode. How to make sure? We would need to check a digital signature issued by the supernode. Private key: Federation name (only known to legit supernodes). Derived public keys would need to be supplied to the edges, yes, as additional command line parameter (strings as generated by the keygen).

We could keep this optional because this kind of attack would involve some knowledge and has some ifs and thens. But if we want to protect against stolen edges, this is part of the full scenario and is not to be neglected.

**Comment by Logan007** [comment]
So, I have evaluated several alternatives (plain ECDH, ECIES for encryption; Ed25519, qDSA for signature). Even though I think that ECIES with its ephemeral public key would be a safer choice in terms of forward secrecy and protecting the transported secret, the dynamic key, (old secrets still protected even if private federation key is leaked), it is the secret itself that can be calculated as soon as the private federation key is known. So, the seemingly better choice does not help in this case, it always melts down to keep our private keys secret...

And as for the signature, if using ECDH, only supernode and that edge will share the secret. Egde will find out if communication was manipulated or tried to be established by a malicious supernode because the shared ECDH secret does not match (leading to wrong key, making communication impossible). So, no signature required yet.

Bottom line: I will use ECDH which will make the federation's public key a must-have parameter at the edge, could be `-P <string>` with _string_ being the federation's public key (derived from federation name, all same at every edge of that federation, keygen to follow).
